src/contractor/contractor.cpp

```diff
@@ -222,36 +226,52 @@ struct SegmentSpeedSource final
 {
     Segment segment;
     SpeedSource speed_source;
+    bool operator<(const SegmentSpeedSource &other)
+    {
+        return std::tie(segment.from, segment.to) > std::tie(other.segment.from, other.segment.to);
+    }
+};
+
+struct Turn final
+{
+    OSMNodeID from, via, to;
+    bool operator==(const Turn &other)
+    {
+        return std::tie(from, via, to) == std::tie(other.from, other.via, other.to);
+    }
 };
 
+struct PenaltySource final
+{
+    double penalty;
+    std::uint8_t source;
+};
+struct TurnPenaltySource final
+{
+    Turn segment;
+    PenaltySource penalty_source;
+    bool operator<(const TurnPenaltySource &other)
+    {
+        return std::tie(segment.from, segment.via, segment.to) >
+               std::tie(other.segment.from, other.segment.via, other.segment.to);
+    }
+};
+using TurnPenaltySourceFlatMap = std::vector<TurnPenaltySource>;
 using SegmentSpeedSourceFlatMap = std::vector<SegmentSpeedSource>;
 
 // Binary Search over a flattened key,val Segment storage
-SegmentSpeedSourceFlatMap::iterator find(SegmentSpeedSourceFlatMap &map, const Segment &key)
+template <typename FlatMap, typename SegmentKey>
+typename FlatMap::iterator find(FlatMap &map, const SegmentKey &key)
```

Comment:
You can just say `auto` as return value and the compiler will pick up the right type